### ISSUE TITLE
Do not initiate breaks when the display is blank/sleeping/off or the screen is locked

### ISSUE THREAD

**Issue by vedgy** [rcaelers/workrave#134]
When the *Reading mode* is on, workrave keeps asking for breaks indefinitely even if there is nothing to read on the screen and the user stopped all mouse/keyboard activity some time ago.

Workrave can check if the screen is readable before initiating a break. If the screen is not readable, check again after some time (a percentage of the shortest interval between enabled breaks perhaps).

For example, on GNU/Linux `xset q` can be used to query the current display state. Unfortunately, it is probably unfeasible for workrave to support all the different screen lockers. But I think some checks (the display state at least) is better than no detection at all.

#120 proposes an alternative manual way to address this issue - by configuring the inactivity timer.

**Comment by vedgy** [comment]
I've implemented a workaround that works perfectly with _xfce4-screensaver_. I added the script below into Autostart list. It has to be constantly running and monitoring D-Bus.

The script always works for screen locking. However it requires the _Activate screensaver when computer is idle_ option to be on in _xfce4-screensaver_ settings in order to detect display blanking/sleeping/off. With this option on the screensaver emits the necessary `ActiveChanged` signal even if the _Enable Screensaver_ option is off.

#### suspend-workrave-on-screensaver-activation
```bash
#!/bin/bash

setWorkraveOperationMode () {
    dbus-send --session --dest=org.workrave.Workrave --type=method_call \
    /org/workrave/Workrave/Core org.workrave.CoreInterface.SetOperationMode \
    string:"$1"
}

justChanged=0

dbus-monitor --session "interface='org.xfce.ScreenSaver',member='ActiveChanged'" |
while read -r line; do  
  if [[ $justChanged == 1 ]]; then
    if [[ "$line" =~ 'true' ]]; then
      setWorkraveOperationMode "suspended"
    elif [[ "$line" =~ 'false' ]]; then
      setWorkraveOperationMode "normal"
    else
      echo "Error! Unexpected activated value: '$line' $(date)"
    fi
    justChanged=0
  elif [[ "$line" =~ 'ActiveChanged' ]]; then
    justChanged=1
  fi
done
```

**Comment by xerus2000** [comment]
This sounds like a great way to implement other custom behavior - is the workrave dbus API documented somewhere?

**Comment by vedgy** [comment]
Not that I know of. You can explore the DBus API in QDBusViewer or a similar tool.

**Comment by xerus2000** [comment]
Btw, maybe you can add a PR to add the script to the contrib section, so it doesn't get lost here :)

**Comment by vedgy** [comment]
The script is now in [contrib](https://github.com/rcaelers/workrave/tree/branch_v1_10/contrib/suspend-workrave-on-screensaver-activation).

**Comment by xerus2000** [comment]
@vedgy nice! Can this be adapted to suspend workrave when there is a fullscreen application running?

**Comment by vedgy** [comment]
Only if your window manager/desktop environment sends a D-Bus signal when an application enters and exits the full-screen mode.

**Comment by xerus2000** [comment]
Well, I now figured out how to check it: https://askubuntu.com/a/1265587
so I guess a cronjob would be the way to go

**Comment by Chealer** [comment]
I did not reproduce this with 1.10.54.

@vedgy: Please indicate at least 1 affected version and which operating systems you experienced this on. It would also be nice to clarify your description (“the screen is not readable”).

**Comment by vedgy** [comment]
Note that this issue is about _Reading mode_ on. In this mode workrave does not track user activity, and it does not check the display state or screen lockers (see the title of this issue). At least this was the case in the latest workrave version _as of 2020_ on Manjaro GNU/Linux. Unless someone implements this feature, it cannot just start working as requested, even after 5 years.

**Comment by Chealer** [comment]
With 1.11 RC 1, I reproduced the screens off case (laptop lid closed, 1 monitor via HDMI turned off).

@vedgy: This is an ITS, so tickets should be titled as such. I suggest:

> Reading mode: Recommends breaks even when screens are sleeping/disabled/off or session locked

**Comment by vedgy** [comment]
I think the current imperative title is suitable for a feature request.

While the issue can be most easily reproduced with the _Reading mode_ on, it might happen with that mode off too. In any case, initiating breaks in the discussed scenarios makes no sense regardless of the state of the _Reading mode_. Though it may be more efficient and just as good in practice to query the display and session states only if the _Reading mode_ is on.

**Comment by Chealer** [comment]
> I think the current imperative title is suitable for a feature request.

Correct. However, as previously indicated, this is an ITS, so tickets should not be requests per se. Tickets should focus on describing 1 issue, and let analysts figure out the best solution(s). It is perfectly fine to offer solutions, but the title should (at least primarily) describe the issue.

Also, the problem is not just the title being imperative. Workrave does not _initiate_ breaks. It _suggests_ (in its terminology, "recommends") taking a break.

> While the issue can be most easily reproduced with the _Reading mode_ on, it might happen with that mode off too.

While that is correct, if you see a concrete scenario where that would be problematic, please explain.

> In any case, initiating breaks in the discussed scenarios makes no sense regardless of the state of the _Reading mode_.

You seem to be assuming that a PC cannot be used without a screen. Yet, many people who are completely blind disprove that.

**Comment by vedgy** [comment]
I don't think your version of the title is much better, and it doesn't fit as well with the issue description, which I don't want to edit to accommodate the title change. I don't think the title matters enough to spend time editing a 6-year-old issue or arguing about it. And by the way, there is an action *Lock Screen* in Xfce, so I wouldn't change that part in any case.

> While that is correct, if you see a concrete scenario where that would be problematic, please explain.

I was thinking about the user typing furiously, then locking the screen and going away, but workrave in the default mode *suggesting* a break a couple of seconds after the locking, i.e. a race condition of sorts.

> You seem to be assuming that a PC cannot be used without a screen. Yet, many people who are completely blind disprove that.

I haven't considered a computer without a screen or with a screen that is normally unused. So you are right: either the proposed feature should be limited to the *Reading mode* or it should be configurable.

**Comment by Chealer** [comment]
> I don't think your version of the title is much better, and it doesn't fit as well with the issue description, which I don't want to edit to accommodate the title change.

The body should be changed anyway (in particular about "asking for breaks indefinitely").

> I don't think the title matters enough to spend time editing a 6-year-old issue or arguing about it. 

I don't think arguing is the best course either, but Workrave was released in 2002, meaning this issue is certainly way older than 6 years, so are you saying you consider that it lost relevance in its lifetime?

> > While that is correct, if you see a concrete scenario where that would be problematic, please explain.
> 
> I was thinking about the user typing furiously, then locking the screen and going away, but workrave in the default mode _suggesting_ a break a couple of seconds after the locking, i.e. a race condition of sorts.

Thanks, but that would be clearer with reproduction steps.

**Comment by vedgy** [comment]
> The body should be changed anyway (in particular about "asking for breaks indefinitely").

The "indefinitely" means that when the *Reading mode* is on Workrave repeatedly suggests breaks while the screen is off or locked, because it never checks the screen readability. 

> so are you saying you consider that it lost relevance in its lifetime?

No, but I think the title and the description are good and understandable enough. Our comments clarify the meaning to excess.

> Thanks, but that would be clearer with reproduction steps.

I don't use the default mode and don't care to test it.

**Comment by Chealer** [comment]
> The "indefinitely" means that when the _Reading mode_ is on Workrave repeatedly suggests breaks while the screen is off or locked, because it never checks the screen readability.

_repeatedly_, yes, but it should stop when there is a rest break.
 
> > so are you saying you consider that it lost relevance in its lifetime?
> 
> No, but I think the title and the description are good and understandable enough. Our comments clarify the meaning to excess.

The main problem is not that the report as a whole is unclear. The problem is Workrave has in the order of magnitude of 1000 reported issues. Tracking an issue involves filtering the list. Without a good title, searches will yield excessive false positives, and more importantly false negatives, in particular with a free ITS engine as basic as this one. And the more unclear titles are, the more time it takes to eliminate "false positives".

> > Thanks, but that would be clearer with reproduction steps.
> 
> I don't use the default mode and don't care to test it.

That's all right, but when it comes to most users, they do not care about and likely even know about the reading mode, so they would rather not spend their time reading an issue report about it. Until we have evidence that the default mode is affected (which should not take long if it is since it is much more used), renaming as suggested would help these users avoid that.
